File: sandboxes/examples/go/bench/perf_lease_exec_delete.go
package main

import (
	"context"
	"fmt"
	"sort"
	"time"

	gocracker "github.com/gocracker/gocracker/sandboxes/sdk/go"
)

func main() {
	ctx := context.Background()
	c := gocracker.NewClient("http://127.0.0.1:9091")
	_ = c.UnregisterPool(ctx, "perfbench-go")
	_, _ = c.RegisterPool(ctx, gocracker.CreatePoolRequest{TemplateID: "perfbench-go", Image: "alpine:3.20", KernelPath: "/home/misael/Desktop/projects/gocracker/artifacts/kernels/gocracker-guest-standard-vmlinux", MinPaused: 8, MaxPaused: 8})
	deadline := time.Now().Add(90 * time.Second)
	for time.Now().Before(deadline) {
		pools, _ := c.ListPools(ctx)
		for _, p := range pools {
			if p.TemplateID == "perfbench-go" && p.Counts["paused"] >= 6 {
				goto ready
			}
		}
		time.Sleep(300 * time.Millisecond)
	}
ready:
	var lease, exec, del []float64
	for i := 0; i < 8; i++ {
		t0 := time.Now()
		sb, err := c.LeaseSandbox(ctx, gocracker.LeaseSandboxRequest{TemplateID: "perfbench-go"})
		if err != nil {
			fmt.Println("lease err:", err); continue
		}
		lease = append(lease, float64(time.Since(t0).Microseconds())/1000)
		t0 = time.Now()
		_, err = sb.Process().Exec(ctx, []string{"echo", "hi"})
		exec = append(exec, float64(time.Since(t0).Microseconds())/1000)
		t0 = time.Now()
		_ = sb.Delete(ctx)
		del = append(del, float64(time.Since(t0).Microseconds())/1000)
	}
	pr := func(name string, xs []float64) {
		sort.Float64s(xs)
		if len(xs) == 0 { fmt.Printf("  %-10s no samples\n", name); return }
		fmt.Printf("  %-10s min=%5.2f  p50=%5.2f  p95=%5.2f  max=%5.2f\n",
			name, xs[0], xs[len(xs)/2], xs[len(xs)-1], xs[len(xs)-1])
	}
	fmt.Println("go-sdk:")
	pr("lease", lease); pr("exec_echo", exec); pr("delete", del)
	_ = c.UnregisterPool(ctx, "perfbench-go")
}

File: sandboxes/examples/python/bench/perf_lease_exec_delete.py
import sys, time, statistics
sys.path.insert(0, '/home/misael/Desktop/projects/gocracker/sandboxes/sdk/python')
from gocracker import Client
c = Client('http://127.0.0.1:9091', timeout=60)
try: c.unregister_pool('perfbench')
except: pass
c.register_pool(template_id='perfbench', image='alpine:3.20', kernel_path='/home/misael/Desktop/projects/gocracker/artifacts/kernels/gocracker-guest-standard-vmlinux', min_paused=8, max_paused=8)
deadline = time.time() + 90
while time.time() < deadline:
    p = [x for x in c.list_pools() if x.template_id == 'perfbench']
    if p and p[0].counts.get('paused', 0) >= 6: break
    time.sleep(0.3)
lease, exec_, delete_ = [], [], []
for _ in range(8):
    t0 = time.perf_counter(); sb = c.lease_sandbox('perfbench'); lease.append((time.perf_counter()-t0)*1000)
    t0 = time.perf_counter(); r = sb.process.exec(['echo','hi'], timeout=5); exec_.append((time.perf_counter()-t0)*1000)
    t0 = time.perf_counter(); c.delete(sb.id); delete_.append((time.perf_counter()-t0)*1000)
def pr(name, xs): xs.sort(); print(f'  {name:10} min={xs[0]:5.2f}  p50={xs[len(xs)//2]:5.2f}  p95={xs[-1]:5.2f}  max={xs[-1]:5.2f}')
print('python-sdk:'); pr('lease', lease); pr('exec_echo', exec_); pr('delete', delete_)
c.unregister_pool('perfbench')
